substrate/frame/support/src/traits/hooks.rs

```diff
@@ -101,21 +101,42 @@ pub trait OnGenesis {
 
 /// See [`Hooks::on_runtime_upgrade`].
 pub trait OnRuntimeUpgrade {
+	/// Advanced lifecycle hook called for all pallets and custom migrations before any other
+	/// [`OnRuntimeUpgrade`] hooks.
+	///
+	/// Sometimes it is useful to have some logic execute prior to any other [`OnRuntimeUpgrade`]
+	/// hooks being called, which [`OnRuntimeUpgrade::before_all`] allows for.
+	///
+	/// When the Executive pallet executes runtime upgrades, hooks are called in this order:
+	/// 1. [`OnRuntimeUpgrade::before_all`] for all custom migrations
```

Comment:
This is technically leaking an implementation detail of executive in the docs here, which is a sign that we are designing things NOT in an intuitive way. 

I don't have a better solution right now though.

Reply:
- I think an API like `fn on_pallet_initialize()` in what I would have expected to see. The default impl of this API would set the storage version. It can be overwritten to something else, and then it is the responsibility of the author to make this happens. 

- Or else, this should be entirely hidden from the user. All pallets set their storage version to the chain on every `on_runtime_upgrade` and this is not configurable. 


My second suggestion is along these lines: I think what you are doing here is in the right direction, but since the API is fairly complicated and counter-intuitive, I think it is best to NOT expose it to the users in any way.

Reply:
I like `fn on_pallet_initialize()`. Unless @bkchr disagrees I'll refactor this PR to move the logic there.

Reply:
Sounds good to me. However, then we probably should change `on_genesis` as this is just a specialization of `on_pallet_initialize()`.

Reply:
WDYT about renaming `on_genesis` to `on_pallet_initialize`?

Reply:
It seems there're a few ways to proceed here, @kianenigma @bkchr I'd appreciate your thoughts:

1. Change the behavior of `on_genesis` to run when the on-chain storage version is unset, regardless of if it is genesis.
2. Create a new hook `on_pallet_initialize` which runs when the on-chain storage version is unset, regardless of if it is genesis.
    a. Also deprecate `on_genesis`, and prompt developers to use `on_pallet_initialize` instead.
    b. Keep `on_genesis`, so developers have the choice of using either hook.

## My thoughts on 1.

I don't like this option very much because
- It changes the behavior of an existing API
- It confuses what the term 'genesis' means

## My thoughts on 2a.

This seems like the best option to me
- Doesn't increase the number of hooks in the API; simpler developer experience
- I struggle to think of a scenario where a developer would want to use `on_genesis` instead of `on_pallet_initialize`

## My thoughts on 2b.

Seems like the way to go if there're indeed scenarios where developers would prefer to use `on_genesis` rather than `on_pallet_initialize`.

Reply:
I also like 2a. So there would be one API, that would typically be implemented like this: 

```
fn on_pallet_initialize(block_number: BlockNumberFor<T>) -> Weight { 
    if block_number.is_zero() { 
        Self::on_genesis() 
    }
    Self::set_storage_version();
} 
```

But then, the question is, what reasonable pallet would implement anything other than the above? and then we would have to go over EVERY pallet, and refactor it to the above code. 

This brings me back to 

> Or else, this should be entirely hidden from the user. **All pallets set their storage version to the chain on every on_runtime_upgrade and this is not configurable**.

And we leave `on_genesis` as-is. 

I think the former is the correct thing to do, but the latter is a much better tradeoff of effort compared to the amount of issues we solve for the time being.

Reply:
> fn on_pallet_initialize(block_number: BlockNumberFor<T>) -> Weight { 
    if block_number.is_zero() { 
        Self::on_genesis() 
    }
    Self::set_storage_version();
} 

I like this :) 

> All pallets set their storage version to the chain on every on_runtime_upgrade and this is not configurable.

This unfortunately doesn't work: 
- `Executive` runs `COnRuntimeUpgrade` migrations *before* the pallet `OnRuntimeUpgrade` hooks
- So, with this method when the `COnRuntimeUpgrade` migrations are executed the versions would still be unset. This is why we need to create a different hook that runs before *all* runtime upgrades.
- Note we can't just reorder to run the pallet `OnRuntimeUpgrade` hooks to run before the `COnRuntimeUpgrade` hooks, because it breaks Basti's `post_upgrade` version consistency checks

I've been thinking a little about this since my last post, and am thinking a good solution for now is to just move the `before_all` (currently implemented in this PR on `OnRuntimeUpgrade` into a trait that only the pallets implement and is hidden from developers. This is a simple approach solves the outstanding issue and doesn't mess with the API. Later, if we chose to, we could make it public and allow developers to extend it like other pallet hooks, also possibly deprecating `on_genesis`. But that is probably an over-optimisation at the moment and doesn't need to be prioritised.

Reply:
> All pallets set their storage version to the chain on every on_runtime_upgrade and this is not configurable.

My woding was bad, it doesn't need to be part of `on_runtime_upgrade` per se, but some similar process. In other words, more or less the current PR is good, just hide it all from users. 

>and am thinking a good solution for now is to just move the before_all (currently implemented in this PR on OnRuntimeUpgrade into a trait that only the pallets implement and is hidden from developers. This is a simple approach solves the outstanding issue and doesn't mess with the API

Exactly.